### ISSUE TITLE
Database: Driver: bad connection error when using MySQL 5.7.18

### ISSUE THREAD

**Issue by marefr** [grafana/grafana#17777]
**What happened**:
Upgraded Grafana from v5.3.4 to v6.2.3 and start seeing the following errors logged:

```
t=2019-06-26T07:00:46+0000 lvl=eror msg="Could not get notification state." logger=alerting.notifier notifier=11192 error="driver: bad connection"
t=2019-06-26T07:00:46+0000 lvl=eror msg="Could not get notification state." logger=alerting.notifier notifier=11192 error="driver: bad connection"
t=2019-06-26T07:00:46+0000 lvl=eror msg="Could not get notification state." logger=alerting.notifier notifier=11192 error="driver: bad connection"
t=2019-06-26T07:01:36+0000 lvl=eror msg="Could not get notification state." logger=alerting.notifier notifier=11192 error="driver: bad connection"
t=2019-06-26T07:02:40+0000 lvl=eror msg="Could not get notification state." logger=alerting.notifier notifier=5312 error="driver: bad connection"
t=2019-06-26T07:02:40+0000 lvl=eror msg="Could not get notification state." logger=alerting.notifier notifier=10802 error="driver: bad connection"
t=2019-06-26T07:02:40+0000 lvl=eror msg="Could not get notification state." logger=alerting.notifier notifier=11192 error="driver: bad connection"
t=2019-06-26T07:02:40+0000 lvl=eror msg="Could not get notification state." logger=alerting.notifier notifier=11192 error="driver: bad connection"
t=2019-06-26T07:02:40+0000 lvl=eror msg="Could not get notification state." logger=alerting.notifier notifier=11192 error="driver: bad connection"
t=2019-06-26T07:02:40+0000 lvl=eror msg="Could not get notification state." logger=alerting.notifier notifier=3561 error="driver: bad connection"
```

**What you expected to happen**:
No errors

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:
Found the following issue which seems related: https://github.com/go-sql-driver/mysql/issues/554
Hints that it's a bug in MySQL. "that’s one of the first (and only) bugs I could find. but as far as I understand from the code there is no `select from select` the actual mysql bug is:
InnoDB: Evaluation of a subquery in a resolving function raised an assertion. (Bug #26909960)"

**Environment**:
- Grafana version: v6.2.3
- OS Grafana is installed on:
- Database is MySQL and the version is 5.7.18

**Comment by marefr** [comment]
Some investigation updates:

The error is logged here: 
https://github.com/grafana/grafana/blob/7cac7435ceb90a52414e718cac70ab61fc60bc77/pkg/services/alerting/notifier.go#L179

Error is returned from somewhere here:
https://github.com/grafana/grafana/blob/7cac7435ceb90a52414e718cac70ab61fc60bc77/pkg/services/sqlstore/alert_notification.go#L456-L502

The working theory is that:
 - There is unique constraint violation error happening here:
https://github.com/grafana/grafana/blob/7cac7435ceb90a52414e718cac70ab61fc60bc77/pkg/services/sqlstore/alert_notification.go#L481
 - Then the [ErrBadConn](https://golang.org/pkg/database/sql/driver/#pkg-variables) error is returned when trying to reuse the existing connection to do one additional query: https://github.com/grafana/grafana/blob/7cac7435ceb90a52414e718cac70ab61fc60bc77/pkg/services/sqlstore/alert_notification.go#L482
 - Basically, the connection goes into a bad state and makes it unusable.

**Comment by marefr** [comment]
This may be related to having a lot of alert rules, like 10000-15000, and having one alert notifier configured as the default.

**Comment by papagian** [comment]
Some additional notes.

The error is generated here:
https://github.com/go-xorm/xorm/blob/master/session_get.go#L101

and the query is:
> SELECT `id`, `org_id`, `version`, `name`, `type`, `access`, `url`, `password`, `user`, `database`, `basic_auth`, `basic_auth_user`, `basic_auth_password`, `with_credentials`, `is_default`, `json_data`, `secure_json_data`, `read_only`, `created`, `updated` FROM `data_source` WHERE `id`=? AND `org_id`=? LIMIT 1

with args:  [1 1]

Still under investigation though.

**Comment by papagian** [comment]
The original error should be this:
https://github.com/go-sql-driver/mysql/blob/master/packets.go#L170

**Comment by papagian** [comment]
The error occurs is `GetOrCreateAlertNotificationState` when tries to start a session for the transaction
https://github.com/grafana/grafana/blob/v6.2.5/pkg/services/sqlstore/alert_notification.go#L457
https://github.com/grafana/grafana/blob/v6.2.5/pkg/services/sqlstore/transactions.go#L65
https://github.com/grafana/grafana/blob/v6.2.5/pkg/services/sqlstore/session.go#L38

In my setup I have reproduced it by having 5000 alert rules and the sql connection is configured with the following values:

- `max_connections`: 1001
- `max_open_conn`: 300
- `max_idle_conn`: 2 (default)

The [error](https://github.com/golang/go/blob/master/src/database/sql/driver/driver.go#L125) seems to happen when the pool has reached the maximum number of open connections and running out of idle connections

>  {MaxOpenConnections:300 OpenConnections:300 InUse:300 Idle:0 WaitCount:3011 WaitDuration:2m42.541541478s MaxIdleClosed:5800 MaxLifetimeClosed:0}

 and therefore may be alleviated with some tuning of the pool size by setting the values of `max_connections`, `max_idle_conn ` and `max_open_conn`.

If in the above setup I change the `max_open_conn` to 0 (default / mean not set) then the specific error does not occur but there are a lot of mysql connection resets:

> grafana_1          | [mysql] 2019/07/18 10:42:45 packets.go:36: read tcp 172.18.0.7:34988->172.18.0.3:3306: read: connection reset by peer

This [article](https://www.alexedwards.net/blog/configuring-sqldb) could be also a good guide.

Furthermore, in Go the number of `maxBadConnRetries` which is checked when trying to get a [connection](https://github.com/golang/go/blob/master/src/database/sql/sql.go#L781) is [hardcoded](https://github.com/golang/go/blob/master/src/database/sql/sql.go#L1390) set to 2. There is also a discussion concerning this [here](https://github.com/golang/go/issues/8834).

**Comment by papagian** [comment]
`max_open_conn` and `max_idle_conn` were default and after consulting with their DBAs I changed it to:
```
max_idle_conn = 400
conn_max_lifetime = 900
max_open_conn = 400
```
but without affecting the error rate. Also, the `max_connections` are 4000.

**Comment by papagian** [comment]
Following this [comment](https://github.com/grafana/grafana/issues/17777#issuecomment-512817459) the error is produced when:

- the maximum number of open connections has been reached
- there are no more idle connections and
- trying to start a session for the transaction.

This happens [here](https://github.com/golang/go/blob/go1.12.5//src/database/sql/sql.go#L1630) because of an open golang [issue](https://github.com/golang/go/issues/32530) that returns an existing connection even though `alwaysNewConn` is set. There is an open [PR](https://github.com/golang/go/pull/32656) for resolving this.

**Comment by papagian** [comment]
The Go [issue](https://github.com/golang/go/issues/32530) has not be resolved yet so it is still around.

**Comment by vbichov** [comment]
@papagian  the underlying issue was solved in go1.15 (please double-check me)
https://github.com/golang/go/issues/32530

but as far as I can see grafana uses go1.14:
https://github.com/grafana/grafana/blob/master/.circleci/config.yml#L41

any plans on updating golang version?

**Comment by aknuds1** [comment]
@vbichov The latest Go version is 1.14.

**Comment by torkelo** [comment]
We are no using Go 1.15 so can we close this?

**Comment by marefr** [comment]
Closed by #27326
